### DISCUSSION TITLE
Alternative suggestion for supporting old and new game mechanics

### DISCUSSION THREAD

**Discussion by cl8n** [ppy/osu#25751] [Ideas] [answered]
the following is a rewrite/expansion of what I said in #general [here](https://discord.com/channels/188630481301012481/1097318920991559880/1183752874933629038).

recently I have been following discussions about scoring and game mechanics relating to lazer's eventual permanent score submission, and I am worried that the general direction is on track to significantly disrupt the timeless nature of osu's maps and leaderboards. there has been so much debate about how mechanics should work, how scoring should work, how Classic Mod should work and what level of parity that can achieve with stable, etc... but in seeing what compromises need to be made, I am increasingly wondering why any of this is really necessary right now.

my opinion essentially boils down to:
**for maps created on stable, lazer should emulate stable's gameplay and scoring.**

and by "emulate stable" I mean as identical as possible, including how hitobjects are judged, how spinners spin, the exact scorev1 algorithm, whatever other baggage is included there.

I will try to explain why I feel this way in this post. I realize that this is kind of "late" to be adding or resurfacing this viewpoint given how much of a pivot it would be from many current efforts...

## improving the game is a good thing

I just want to get this out of the way to avoid confusing what I'm trying to say: most of the changes I have seen made to gameplay/scoring seem well thought out, more intuitive, more beginner-friendly, overall just solid improvements. and it's great that they are getting active feedback from testers and developers alike. I'm happy to see these changes being made and I'm not trying to say that they should be thrown away.

## ...but improving the game doesn't have to compromise what's already here

the reality is, though, that for many, the entire appeal of playing osu online hinges on how it has specifically avoided any kind of upheaval of leaderboards or drastic changes to game mechanics. each ranked map has one continuous leaderboard where players compete with one set of rules, and it's more or less been that way since 2008 or so, minus some accidents. I feel strongly that if this quality were ever lost, it would be alienating to the degree of not feeling like the same game anymore.

so I worry when I read the talks lately about how things are shaping up for lazer's "release". it seems almost definitive that lazer will use a new scoring algorithm, and that stable scores will have to perform some lossy conversion if shown alongside lazer scores. it seems almost definitive that lazer will modify stable's mechanics by default, only providing something of a compatibility layer via Classic Mod, which still won't produce the same results as if you had played on stable. some people are in the camp of trying to minimize issues with merged leaderboards, while others argue to just keep them separate.

I find it difficult to be satisfied with any of the stances people are taking here. there's always *something* that is being compromised, even if in a clever way. there's always *something* that effectively fragments the system between a "new" and "old" set of scores for each map.

even beyond leaderboards, I find it concerning that the default game mechanics -- most notably, accuracy requirements for slider heads -- will change for maps that have been created under different rules. I won't elaborate much here because I see this point has come up in past community meetings, but it's just another thing to be dissatisfied with, adding to the above.

## my proposal

my general wish here is for improving the game's future to not come at the cost of changing existing maps and leaderboards. so this is my idea:

- maps created on stable (file format <=14 i guess?) should exactly match stable's mechanics and scoring
- maps created on lazer can change whatever you see fit and begin to establish new and improved mechanics and scoring

it can be thought of as shifting the viewpoint from thinking of certain scores as "classic" to thinking of certain *maps* as "classic".

this would not necessarily restrict the "classic" maps from ever supporting anything new. for example, there are probably many new mods that could easily function on classic maps without really disrupting the leaderboards or anything. the implication is just that changes to what happens on classic maps should be carefully applied and heavily justified, as have been changes to ranked maps in the past, such as tweaks to HP drain and spinner logic.

this would let non-"classic" maps be much more free to apply new scoring systems and game mechanics without comparing everything to the baggage of 16 years that came before it.

I have many further thoughts about this topic but I'll keep it at that for now thx for reading my wall of text

**Comment by Zyfarok** [comment]
Totally disagree here. The huge majority of maps benefit from the improved lazer mechanics (which mostly involves sliders) and improved scoring, thus the DEFAULT (=main) leaderboard should include scores played with those new mechanics. This includes slider-head accuracy which has always been seen as "how it should have been" to anyone with enough ScoreV2 experience and the amount of maps where it doesn't work well is extremely small.

Other than that, whatever is done to stable scores (separate leaderboard or not, ability to filter leaderboards to only see those, etc...) I couldn't care less as long as they are not completely lost, that they can be seen "in isolation" (a filter is enough) and that they still award pp.

**Reply by cl8n** [reply]
my argument is not so much about whether the maps "benefit" from significant changes or not, rather that there are always significant downsides that come along with that. but if your opinion is just "I couldn't care less", then I think that is a fundamental disagreement we probably can't convince each other of -- the thing you don't care about is probably one of my top priorities in this game, just not one I often have to consciously recognize since it isn't often challenged

**Reply by Zyfarok** [reply]
The upside of lazer mechanics is huge and the downside is only relevant in the short term. In 5 years from now, there will probably be more lazer score than all stable scores ever and the skillcaps be pushed even further. By that time people will care much less about stable scores.

Also, locking all previous maps to stable mechanics just means we will need to wait month if not year before we can actually play a wide variety of maps with the new lazer mechanics which is stupid. Then there's the question of what do we do if someone want to rerank the same map just to have it be playable with lazer mechanics ? At this point this will show how stupid it would be to not simply have it by default on all maps without the need to resubmit them.

**Reply by cl8n** [reply]
as said in my last reply I think these are just disagreements in priorities; I do not agree that these issues become less relevant over time. my argument is specifically about preserving a continuity of how leaderboards work and it will never become less concerning than it is right now to me if old scores are playing by different rules than new ones.

likewise I don't think it's "stupid" to have people "wait" for new mechanics because from the perspective of this argument it should never have been assumed that these mechanics can change on ranked leaderboards to begin with...

the resubmitting issue seems valid but I don't really understand how you jump from "someone wants [...] to have it be playable with lazer mechanics" to "have it by default on all maps". like, what about anyone who *doesn't* want the new mechanics to be the default on their already-ranked map? these are sort of mirrored problems that stem from having different rankable mechanics available *at all*, so I don't know if there is a good solution there, regardless of how this all goes down

**Comment by WitherFlower** [comment]
> this would not necessarily restrict the "classic" maps from ever supporting anything new

Given how intact you want things to be preserved, I doubt any new significant mechanic will ever be available on old maps.

Now, if I can't switch to lazer and play all the maps I know and love with custom rate changes, a better scoring algorithm, and better gameplay mechanics (no notelock, slider accuracy, etc...), what's even the point of switching to the new client ?

**Reply by cl8n** [reply]
> Given how intact you want things to be preserved, I doubt any new significant mechanic will ever be available on old maps.

yeah, that is essentially what I'm suggesting should happen. "new significant mechanics" would not and have not been introduced in stable since around 2008, for good reason, and I do not see why this update in particular should allow itself to discard those values.

> Now, if I can't switch to lazer and play all the maps I know and love with custom rate changes, a better scoring algorithm, and better gameplay mechanics (no notelock, slider accuracy, etc...), what's even the point of switching to the new client ?

I'm not against options to try new mechs on old maps, just as there is the "V2" mod in stable today. it just shouldn't be messing with the leaderboard or forcing itself as the default, according to this proposal. maybe I should have written taht in OP as well lol i just thought it was getting too long...

and uh, I can't tell if that question was plainly rhetoric or not but just in case, the point of switching to the new client for me at least is to get all of the changes that *aren't* game mechanics or scoring changes... which is pretty much the entire thing

**Reply by WitherFlower** [reply]
> "new significant mechanics" would not and have not been introduced in stable since around 2008

Well that's true, and I'd argue it's about time we got a breath of fresh air

> just as there is the "V2" mod in stable today

I don't play with scoreV2 on stable even though it's been here for the entire 6 years I've been playing the game. The reason why I don't play with scoreV2 is that it's unranked and makes all my scores pointless. (there's also the 1m cap but that's since been resolved in lazer so i digress.) It's also one of the reasons I don't do tournaments anymore, cause all the good scores I get during those are just left alone in my local rankings which I never look at.

> get all of the changes that aren't game mechanics or scoring changes... which is pretty much the entire thing

Well if that's that, then I'll probably be sticking with the client that has a more usable song select, and skinnability on more visual elements. Sure, the editor has gotten some pretty cool updates, but I'm not a mapper, so I don't really care about it. Maybe I'll just hop on lazer for playlists from time to time, but that's about all that lazer has in terms of interesting features for me once you remove the new gameplay elements, because to me, that's the actual point of "the entire thing".

**Comment by Digitalfear117** [comment]
I agree very much with the sentiment expressed here. For me, the history of osu!, and the idea that all players that play a map, play under the same rules is really important. It’s what draws me most into the game, the idea that plays and players can have a legacy that isn’t just pp, which is constantly shifting and being tweaked. At least the leaderboard will always exist, only changing as new players set more incredible scores under the same rules. But that’s my perspective as someone who feels very strongly about the leaderboard of maps. 

I think the issue here is that this ideal doesn’t really align with the average players, and I think this can be seen from previous survey’s and top player’s comments. Does the average player really care? Or do they see lazer as an opportunity to gain more pp as they can now rate edit DT to a precise amount, or use DA to increase OD. Or correct the “mistake” of not having slider accuracy? Or maybe players want osu! to feel as it’s evolving in gameplay? I’m not entirely sure really, but it seems players want a different direction from my limited understanding.

**Comment by longnguyen2004** [comment]
There are also problems from the technical side. The dev team has stated multiple times that replicating stable 1:1 will have implications in many parts of the codebase. I need to reiterate: This is a rewrite from scratch, not a simple UI upgrade of the game. There's a reason why it took so long, and why it's not called osu!next (see [this blog post](https://blog.ppy.sh/a-definitive-lazer-faq/) for more details).

Bringing back 16 years of legacy code with various hacks into this new framework that they've rebuilt and futureproof is going to take a lot of work, and completely defeats the purpose of the rewrite, not to mention the other 3 gamemodes, with their own set of quirks and weird behaviors (especially osu!mania, see [this](https://niv.gay/osumania-research-lab/index.html)). The dev team doesn't expect complete backwards compatibility, and I'm pretty sure the average player doesn't either. As long as things are mostly where they should be (no weird stuff like a 6-digit sniping top players for example), then it should be fine.

(Of course, I'm not a dev in the team, and I don't have access to stable's code, but when the lead developer itself is frustrated with the codebase, to the point of doing a complete rewrite, then maybe we should trust them a bit)

**Reply by bdach** [reply]
> The dev team doesn't expect complete backwards compatibility

Yes. We have foregone complete backwards compatibility where it is not feasible.

For instance, it is currently the case that there are beatmaps that do not decode from the `.osu` file in the same way that stable does this. Most often this is caused by *differences in behaviour of floating-point arithmetic between .NET Framework and .NET (Core)*. Fixing this would require us to *port the exact implementation of floating-point arithmetic to .NET (Core)*. Somehow.

We have tried and still continue to try to get as closely to the original as possible, but still, in numerous instances we just threw up our hands and let it go. Most recent example: https://github.com/ppy/osu/pull/25725. Then see comments such as

https://github.com/ppy/osu/blob/15ce45ac71e3e561fb56870141df15c317794122/osu.Game.Rulesets.Taiko/Beatmaps/TaikoBeatmapConverter.cs#L177-L179

We have also foregone - wherever possible - porting _literal stable bugs_. There are several mechanic-impacting _literal bugs_ in stable that have not been ported because in our opinion doing so would be just stupid (see [example](https://discord.com/channels/188630481301012481/188630652340404224/1144518158158397501)).

And then we ported some anyway. See https://github.com/ppy/osu/blob/master/osu.Game/Skinning/LegacyColourCompatibility.cs for an example.

**There is no 1:1 porting of stable, warts and all, without extraordinary and unreasonable engineering effort.** There will be changes. At this point the question would be where the line of "acceptable change" is drawn. But that's not what the OP wants, seemingly.

Moreover:

> * maps created on stable (file format <=14 i guess?) should exactly match stable's mechanics and scoring
> * maps created on lazer can change whatever you see fit and begin to establish new and improved mechanics and scoring

This would likely require us to maintain **two codebases at once** to maintain a healthy pace of forward progress.

**Reply by Zyfarok** [reply]
@bdach I don't want to waste any of your time so you can ignore if you want, but couldn't we have a tool based on stable code-base that would to convert beatmaps and replays to a format that is easily usable by lazer and not subject to floating point arithmetic issues ?
To at least make sure stable replays are 1:1 with stable.
(Maybe it's already crossing the line of "unreasonable engineering effort" idk)

**Reply by bdach** [reply]
A crude approximation of that _for beatmaps_ exists in cutting edge and it's called [conversion mappings](https://github.com/ppy/osu/wiki/Generating-osu!stable-conversion-mappings). We already use that to validate beatmap conversion results. But that is still tied to a desktop client that needs .NET Framework to run. Sustained operation of this setup requires a significant amount of infrastructure, and probably tearing out the conversion code to something that can run on a server rather than be a game instance jury-rigged to someone's desktop. This is weeks to months of development time and then requires sustained maintenance.

I haven't even _considered_ what this proposed new format means for stuff like beatmap mirrors etc. It is probably a doubling of required storage space, if not worse.

As for replays... I don't even know where to begin with this one. Our replay setup relies on game mechanics being reproduced in real time. If the replays are to play back exactly, all mechanics must match to a T. See: https://github.com/ppy/osu/issues/11311 etc. Any other approach would require somehow reengineering the entire concept of replay playback to not store the user's inputs, but the _entire game state_ instead. I am not willing to even consider what this would look like.

**Reply by Zyfarok** [reply]
I feel like there's no need to store the entire game-state as all that matters is the judgement results (maybe the score updates if recomputing it can cause issues, and the spinner half-turns and the slider-ball activation/deactivation for matched visuals) but I get your point.

**Comment by Flamiii** [comment]
I understand your concern, because there absolutely are going to be many maps where the leaderboards will probably be dominated by osu!(lazer) scores because the map with osu!(lazer) gameplay mechanics is easier than it would be on osu!stable. But at the same time, that's just something we have to accept in the name of making forward progress. There will likely never be a better time to make changes to gameplay mechanics and scoring than now, and I think it would be foolish to not do anything for the sake of preservation when osu!stable has some big flaws. For example, many players dislike osu!stable notelock, and many also find it weird that a supposed "rhythm game" doesn't have slider accuracy. I think making improvements like this are important because the game's lifeblood comes down to newer players. osu!(lazer) is likely going to cause the game to explode in popularity, and it's important that those newer players trying out the game for the first time have a good experience and play the best version of the game possible.

**Comment by TextAdventurer12** [comment]
Isn't this the purpose of the classic mod?

**Comment by peppy** [comment, marked as answer]
My answer to this is to use stable if you want this. Lazer is our vow to keep the game moving forward, bringing new experiences and fixing issues that have been around for ages. Gimping existing maps to old mechanics and strictly barring playing with new features feels like really bad move for the community, especially for new players who will largely benefit and come to enjoy the new mechanics.

Stable will remain, untouched for the most part. We will only port back mechanics where they make sense.

One thing we have considered is allowing certain beatmaps to force classic mod on. That's still a direction we can explore, but the hope is that we don't need to resort to this (or we leave it to users to decide themselves).

**Reply by cl8n** [reply]
"use stable" doesn't solve the problem for me but I think you understand that too.

having new scores not play exactly fair with old ones on the same leaderboards makes it feel to me as if all of those leaderboards are starting over in a way. which is just something I never thought would be considered, given the decade of experience I have seeing, if anything, explicit efforts to *preserve* the continuity and fair play across time on ranked leaderboards

I don't disagree about the existence of old issues or potentially worse experience on a lot of maps compared to if they were using new mechanics. my perspective is just that I think those past mistakes are set in stone for various reasons and should have been treated as such. if your mind can't be changed on that point then okay and that is end of thread I guess. but keep in mind that the perspective like mine more or less comes from seeing it respected in practice for so long; I'm not the only person that developed this expectation for a kind of "permanent", ongoing, unchanging competition on the ranked maps of the past

**Reply by longnguyen2004** [reply]
There's also a not insignificant portion of people who are 100% in favor of the new mechanics, and an even bigger portion of people who either just don't care or are happy with a middle ground. Trying to match stable 1:1 will cause extra maintenance burden, and also disappoint the people who are waiting for some kind of changes to this game. The goal is to piss off the least number of people, and that is what the dev team has been doing for game balance.

Anyway, in the first period of ranked play, stable leaderboard won't contain lazer leaderboard, so the leaderboard would be exactly the same. That would be enough for you I think?
